src/wallet/export/caravan.rs

```diff
@@ -0,0 +1,461 @@
+use crate::error::Error;
+use crate::keys::{DerivableKey, DescriptorKey, SortedMultiVec};
+use crate::miniscript::descriptor::{ShInner, WshInner};
+use crate::miniscript::MiniscriptKey;
+
+/// Structure that contains the export of a wallet
+///
+/// For a usage example see [this module](crate::wallet::export::caravan)'s documentation.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CaravanExport {
+    /// Vault name
+    pub name: String,
+    /// Caravan address type,
+    #[serde(rename = "addressType")]
+    pub address_type: CaravanAddressType,
+    /// Caravan network
+    network: CaravanNetwork,
+    /// Caravan client
+    pub client: CaravanClient,
+    /// Signing quorum
+    pub quorum: Quorum,
+    /// Extended public keys
+    #[serde(rename = "extendedPublicKeys")]
+    pub extended_public_keys: Vec<CaravanExtendedPublicKey>,
+    /// Starting address index, always 0 when exporting
+    #[serde(rename = "startingAddressIndex")]
+    pub starting_address_index: u32,
+}
+
+impl CaravanExport {
+    /// Get the bitcoin network value
+    pub fn network(&self) -> Network {
+        match self.network {
+            CaravanNetwork::Mainnet => Network::Bitcoin,
+            CaravanNetwork::Testnet => Network::Testnet,
+        }
+    }
+    /// Get the descriptor value
+    pub fn descriptor(&self) -> Result<Descriptor<DescriptorPublicKey>, Error> {
+        let required = self.quorum.required_signers;
+        let network: Network = self.network();
+
+        let result = match self.address_type {
+            CaravanAddressType::P2sh => {
+                let keys: Vec<DescriptorKey<Legacy>> = self.descriptor_keys()?;
+                descriptor! { sh ( sortedmulti_vec(required, keys) ) }
+            }
+            CaravanAddressType::P2shP2wsh => {
+                let keys: Vec<DescriptorKey<Segwitv0>> = self.descriptor_keys()?;
+                descriptor! { sh ( wsh ( sortedmulti_vec(required, keys) ) ) }
+            }
+            CaravanAddressType::P2wsh => {
+                let keys: Vec<DescriptorKey<Segwitv0>> = self.descriptor_keys()?;
+                descriptor! { wsh ( sortedmulti_vec(required, keys) ) }
+            }
+        }
+        .map_err(|e| Error::Descriptor(e));
+
+        match result {
+            Ok((d, _, n)) => {
+                if n.contains(&network) {
+                    Ok(d)
+                } else {
+                    Err(Error::InvalidNetwork {
+                        requested: network,
+                        found: *n.iter().last().expect("network"),
+                    })
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn descriptor_keys<Ctx: ScriptContext>(
+        &self,
+    ) -> Result<Vec<DescriptorKey<Ctx>>, DescriptorError> {
+        let result = self
+            .extended_public_keys
+            .iter()
+            .map(|k| {
+                let fingerprint = k.xfp;
+                let key_path = k.clone().bip32_path;
```

Comment:
If you only need the path it's slightly faster to only clone that instead of the whole struct and then throwing away part of it.

```suggestion
                let key_path = k.bip32_path.clone();
```